### ISSUE TITLE
An empty ring should be an empty input!

### ISSUE THREAD

**Issue by brianharvey** [jmoenig/Snap--Build-Your-Own-Blocks#1204]
This behavior is wrong:

![google chrome001](https://cloud.githubusercontent.com/assets/3885074/14993838/9f22a722-1121-11e6-97fb-5f12f7422f1e.png)

I know you think the ring is a non-empty occupier of the input slot, but it isn't.  And it's messing up the ability of the TEST block to test HOFs: [https://piazza.com/class/i7hy3udaukx6su?cid=211](url)

**Comment by jmoenig** [comment]
NO, NO, and NO!!!
I am not ever going to give in to such kludges. A ring is a ring, it is not ever an empty input. You cannot call an empty ring with another non-empty one and expect something magic to happen, because that's totally stupid. I don't care what teachers think, because you're making them expect wrong things that you want to have yourself. This is very upsetting.

**Comment by jmoenig** [comment]
I also disagree with your MAP block

**Comment by jmoenig** [comment]
This is so full of shit it makes me weep.

**Comment by brianharvey** [comment]
Yeah, but what do you *really* think?

How would you like people to achieve the desired result here?

**Comment by jmoenig** [comment]
![test](https://cloud.githubusercontent.com/assets/2596285/14995325/818f794e-1174-11e6-9943-efebd6885c57.png)

**Comment by brianharvey** [comment]
Sigh, yeah, that's what I told Emil to do, but the first time I got wrong which function needs explicit parameters; I thought it was LETTER and LENGTH rather than TEST.  And if *I* can't understand it, I don't think it's reasonable to expect users to understand where they need formal parameters and why.

Maybe to make you happy we should make the input notation for procedure types be a dashed-line ring or something.  Then if you drag in a non-procedure-or-variable, it automatically get ringified.  And then if you drag out the ring, the dashed line becomes visible again.  (But if you drag out the function inside the ring, the ring stays.  I think.  Or maybe not.)

**Comment by jmoenig** [comment]
There is nothing confusing whatsoever about replacing the ring with a formal parameter for a ring. What *would* be confusing is *not* replacing it, because calling an empty ring with a non-empty one somehow causes the inner contents of the non-empty ring to end up inside the first empty ring, or *replacing* the empty ring with the non-empty one. And if you'd just put formal parameters everywhere it would also totally have worked.

**Comment by brianharvey** [comment]
You are looking at it from the implementor's point of view.  Every user except you thinks that an unfilled slot of type procedure is just like an unfilled slot of any other type.

But I think we're going to have to argue this f2f.  After I've gotten a few beers into you.

**Comment by jmoenig** [comment]
I think you're looking at it from the beer's view :-)
You can't seriously be advocating that

```JavaScript
((function () {})(function (n) {return n * 2}))(5)
````

should return ```10```, can you?

**Comment by brianharvey** [comment]
Javascript doesn't have Snap<i>!</i>'s empty-slot-filling feature for non-procedure inputs either.
`(function () {return ()+3;})(5)`
doesn't return 8.  But if it did, then I would, yeah.

**Comment by brianharvey** [comment]
To me, your position is exactly analogous to if you were claiming that the little picture of a list with two empty items were an *actual list* with two empty items, and therefore we shouldn't do substitution into list slots.

**Comment by jmoenig** [comment]
... which is indeed what I'm claiming for the list primitive:

![list](https://cloud.githubusercontent.com/assets/2596285/15020479/7d477c84-1222-11e6-9815-048ca6b83045.png)

**Comment by brianharvey** [comment]
No, I'm not talking about issues of variadicness, although that's another issue we should discuss.  I mean just something like
![listmap](https://cloud.githubusercontent.com/assets/3885074/15021042/79694fb4-11d9-11e6-8775-4eb28a2c792c.png)

Long ago we had a discussion about how to represent procedures  in variable watchers, and you pointed out correctly that technically the value is just the procedure, not a niladic procedure that returns the procedure, and so there shouldn't be a grey ring.  But we ended up agreeing that users will see the grey ring as meaning that what they're looking at is a procedure -- it functions for them as a type tag, not as a niladic procedure.  And it's the same thing for the ring in a higher order block: It's a reminder that a procedure goes there, just as ![list-slot](https://cloud.githubusercontent.com/assets/3885074/15021184/3d2b68c4-11da-11e6-84b7-60047381a29e.png) is a reminder that a list goes there, not an actual two-item list.

**Comment by cycomachead** [comment]
> But I think we're going to have to argue this f2f. After I've gotten a few beers into you.

Didn't we basically have this argument last november in Brian's office? :P

**Comment by brianharvey** [comment]
We've definitely had it before, I'm not sure where.  Do you remember who won last time? :-)

**Comment by cycomachead** [comment]
> We've definitely had it before, I'm not sure where. Do you remember who won last time? :-)

Ooooh, can I pick my side? 

I think no one came to a conclusion, but I think we decided that we needed more info.

**Comment by joefarebrother** [comment]
I have to agree with Brian here. In any other context, an empty input slot is replaced with a formal parameter. 

To me, an empty ring should be the identity function.
